pkg/image/registry/imagestream/etcd/image.go

```diff
@@ -0,0 +1,186 @@
+package etcd
+
+import (
+	"fmt"
+
+	metainternalversion "k8s.io/apimachinery/pkg/apis/meta/internalversion"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/watch"
+	apirequest "k8s.io/apiserver/pkg/endpoints/request"
+	"k8s.io/apiserver/pkg/registry/rest"
+	"k8s.io/client-go/tools/cache"
+
+	imageapi "github.com/openshift/origin/pkg/image/apis/image"
+)
+
+// ImageLayerIndex is a cache of image digests to the layers they contain.
+// Because a very large number of images can exist on a cluster, we only
+// hold in memory a small subset of the full image object.
+type ImageLayerIndex interface {
+	cache.SharedIndexInformer
+}
+
+type ImageStore interface {
+	rest.Watcher
+	rest.Lister
+}
+
+// NewImageLayerIndex creates a new index over a store that must return
+// images.
+func NewImageLayerIndex(store ImageStore) ImageLayerIndex {
+	ctx := apirequest.NewContext()
+	informer := cache.NewSharedIndexInformer(&cache.ListWatch{
+		ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+			obj, err := store.List(ctx, &metainternalversion.ListOptions{
+				ResourceVersion: options.ResourceVersion,
+				Limit:           options.Limit,
+				Continue:        options.Continue,
+			})
+			if err != nil {
+				return nil, err
+			}
+			list, ok := obj.(*imageapi.ImageList)
+			if !ok {
+				return nil, fmt.Errorf("unexpected store type %T for layer index", obj)
+			}
+			// reduce the full image list to a smaller subset.
+			out := &metainternalversion.List{
+				Items: make([]runtime.Object, len(list.Items)),
+			}
+			out.Continue = list.Continue
+			out.ResourceVersion = list.ResourceVersion
+			for i, image := range list.Items {
+				out.Items[i] = &ImageLayers{
+					Name:     image.Name,
+					Layers:   image.DockerImageLayers,
```

Comment:
These layers have different order in different images. ImageAPI reorder them and add `imageapi.DockerImageLayersOrderAnnotation`, but this code drop this infomation. After that, these layers are copied as is, but without an annotation indicating the order.

Also there is no `imageapi.ManagedByOpenShiftAnnotation` for image. Therefore, then the registry does not have the ability to understand this.

Reply:
I’ll reorder them before returning.

With regards to pullthrough off, I think we probably should drop support for disabling it.  We will depend on it for secured images (when registry access for redhat is made protected), and local resolution depends on it.  Is there a good use case for continuing to support it?

Reply:
> Is there a good use case for continuing to support it?

You tell me :) We always had the option to turn on/off pullthrough. Because of this, the registry's code is more complex than it could be because we always need to remember the type of the image and sometimes pretend that we do not know about it.

If you think that we can drop it, then let's do it. Please open an issue to remove it.

But right now without `imageapi.ManagedByOpenShiftAnnotation` pullthrough will be broken.

Reply:
Unfortunately, we can not completely remove `imageapi.ManagedByOpenShiftAnnotation`.

Reply:
Let me chat with Ben - if we think we're dependent on pull through now for protected image access, then I think it's time to remove the support for pull through false.  Will respond back today.

Do we need the annotation for anything else specific to the use of this call (in the place where we check whether an image stream has "access" to a layer)?

A long time ago we talked about changing the meaning of the annotation (or replacing it with a new annotation) which was:

> The person who imported this image metadata has proven they have access to each of the layers in the image they imported.

A background controller (the mirror controller) would not set the annotation until it had either pulled all layers from the source (in the background) or otherwise verified the source had those layers (check that the source can serve the contents).  Before the annotation was set, only pull through would work, and users would be blocked from tagging the image into another image stream.  Once the annotation was set, the image was now effectively managed (we've imported the image and the metadata).  The downside of this is that it forces mirroring which can be expensive (today mirroring is mostly controlled by whether you set ReferencePolicy in practice).

Reply:
> Let me chat with Ben - if we think we're dependent on pull through now for protected image access, then I think it's time to remove the support for pull through false.

OCP will be, Origin won't be.  Perhaps that's sufficient though.

(Technically someone could also setup an OCP cluster that got protected images from elsewhere like an internal mirror, so it is possible to run an OCP cluster w/o pullthrough, also).

But i'm certainly a fan of simplifying.

Reply:
I think we closed on this, right?  Dropping pullthrough: false support?

Reply:
yeah, being removed here: https://github.com/openshift/image-registry/pull/103

i think it's currently stuck on that test that needs the docker socket for which i have a PR open to mark as local.